{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8fe64f70",
   "metadata": {},
   "source": [
    "# Chapter ‍18 KEGG"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7522af3e",
   "metadata": {},
   "source": [
    "KEGG (http://www.kegg.jp/) is a database resource for understanding high-level functions and utilities of the biological system, such as the cell, the organism and the ecosystem, from molecular-level information, especially large-scale molecular datasets generated by genome sequencing and other high-throughput experimental technologies.\n",
    "\n",
    "Please note that the KEGG parser implementation in Biopython is incomplete. While the KEGG website indicates many flat file formats, only parsers and writers for compound, enzyme, and map are currently implemented. However, a generic parser is implemented to handle the other formats."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e481a9e6",
   "metadata": {},
   "source": [
    "## 18.1 Parsing KEGG records\n",
    "Parsing a KEGG record is as simple as using any other file format parser in Biopython. (Before running the\n",
    "following codes, please open http://rest.kegg.jp/get/ec:5.4.2.2 with your web browser and save it as\n",
    "`ec_5.4.2.2.txt`.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c8b2c02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Isomerases;',\n",
       " 'Intramolecular transferases;',\n",
       " 'Phosphotransferases (phosphomutases)']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.KEGG import Enzyme\n",
    "records = Enzyme.parse(open(\"ec_5.4.2.2.txt\"))\n",
    "record = list(records)[0]\n",
    "record.classname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d0ccacdc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'5.4.2.2'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record.entry"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0587654b",
   "metadata": {},
   "source": [
    "## 18.2 Querying the KEGG API"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d362a78e",
   "metadata": {},
   "source": [
    "Biopython has full support for the querying of the KEGG api. Querying all KEGG endpoints are supported; all methods documented by KEGG (http://www.kegg.jp/kegg/rest/keggapi.html) are supported. The interface has some validation of queries which follow rules defined on the KEGG site. However, invalid queries which return a 400 or 404 must be handled by the user."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ec16e39f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Isomerases;',\n",
       " 'Intramolecular transferases;',\n",
       " 'Phosphotransferases (phosphomutases)']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.KEGG import REST\n",
    "from Bio.KEGG import Enzyme\n",
    "request = REST.kegg_get(\"ec:5.4.2.2\")\n",
    "open(\"ec_5.4.2.2.txt\", \"w\").write(request.read())\n",
    "records = Enzyme.parse(open(\"ec_5.4.2.2.txt\"))\n",
    "record = list(records)[0]\n",
    "record.classname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6a00d959",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'5.4.2.2'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " record.entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c50a5692",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import REST\n",
    "human_pathways = REST.kegg_list(\"pathway\", \"hsa\").read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dd892777",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter all human pathways for repair pathways\n",
    "repair_pathways = []\n",
    "for line in human_pathways.rstrip().split(\"\\n\"):\n",
    "    entry, description = line.split(\"\\t\")\n",
    "    if \"repair\" in description:\n",
    "        repair_pathways.append(entry)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "af91eeb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 3 repair pathways and 100 repair genes. The genes are:\n",
      "OGG1, NTHL1, NEIL1, NEIL2, NEIL3, UNG, SMUG1, MUTYH, MPG, MBD4, TDG, APEX1, PNKP, TDP1, POLB, POLL, HMGB1, PARP1, PARP2, PARP3, PARP4, PARG, ADPRS, APTX, XRCC1, POLG, POLG2, LIG3, POLD1, POLD2, POLD3, POLD4, POLE, POLE2, POLE3, POLE4, PCNA, RFC1, RFC4, RFC2, RFC5, RFC3, FEN1, LIG1, RBX1, CUL4B, CUL4A, DDB1, DDB2, XPC, RAD23B, RAD23A, CETN2, ERCC8, ERCC6, UVSSA, POLR2A, POLR2B, POLR2C, POLR2D, POLR2E, POLR2F, POLR2G, POLR2H, POLR2I, POLR2L, POLR2K, POLR2J, POLR2J3, POLR2J2, POLR2M, CDK7, MNAT1, CCNH, ERCC3, ERCC2, GTF2H5, GTF2H1, GTF2H2, GTF2H2C_2, GTF2H2C, GTF2H3, GTF2H4, ERCC5, BIVM-ERCC5, XPA, RPA1, RPA2, RPA3, RPA4, ERCC4, ERCC1, SSBP1, PMS2, MLH1, MSH6, MSH2, MSH3, MLH3, EXO1\n"
     ]
    }
   ],
   "source": [
    "# Get the genes for pathways and add them to a list\n",
    "repair_genes = [] \n",
    "for pathway in repair_pathways:\n",
    "    pathway_file = REST.kegg_get(pathway).read()  # query and read each pathway\n",
    "\n",
    "    # iterate through each KEGG pathway file, keeping track of which section\n",
    "    # of the file we're in, only read the gene in each pathway\n",
    "    current_section = None\n",
    "    for line in pathway_file.rstrip().split(\"\\n\"):\n",
    "        section = line[:12].strip()  # section names are within 12 columns\n",
    "        if not section == \"\":\n",
    "            current_section = section\n",
    "        \n",
    "        if current_section == \"GENE\":\n",
    "            gene_identifiers, gene_description = line[12:].split(\"; \")\n",
    "            gene_id, gene_symbol = gene_identifiers.split()\n",
    "\n",
    "            if not gene_symbol in repair_genes:\n",
    "                repair_genes.append(gene_symbol)\n",
    "\n",
    "print(\"There are %d repair pathways and %d repair genes. The genes are:\" % \\\n",
    "        (len(repair_pathways), len(repair_genes)))\n",
    "print(\", \".join(repair_genes))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
